pkg/cache/v3/snapshot.go

```diff
@@ -103,6 +130,11 @@ func NewSnapshotWithResources(version string, resources SnapshotResources) Snaps
 	out.Resources[types.Runtime] = NewResources(version, resources.Runtimes)
 	out.Resources[types.Secret] = NewResources(version, resources.Secrets)
 	out.Resources[types.ExtensionConfig] = NewResources(version, resources.ExtensionConfigs)
+
+	// Initialize the snapshot version map here
+	v, _ := initializeVMap(&out)
```

Comment:
I would probably just do the initialization of the nested structures during `ConstructVersionMap`, this seems very decoupled from where it gets used, and we end up doing useless work in case the snapshot is never used with delta

Reply:
Yup makes sense

Reply:
@snowp so it looks like this is what caused the test to blow up. When I move this into the `ConstructVersionMap` function, it somehow always overwrites and then the watches don't trigger properly. In the test we don't make a new snapshot twice we just overwrite the pre-existing snapshot. I even tried doing something like:
```go
	if len(s.VersionMap) == 0 {
		// initialize ...
		v, err := initializeVMap(s)
		if err != nil {
			return fmt.Errorf("failed to initialize verison map: %v", err)
		}
		s.VersionMap = v
	}
```
in ConstructVersionMap but that didn't work. I'm perplexed by this one